### PR TITLE
bluetooth: drivers: Check if context is atomic before obtaining lock

### PR DESCRIPTION
**Pull request by rugeGerritsen**
According to https://github.com/zephyrproject-rtos/zephyr/issues/16273
it is necessary to check if the context is necessary to check if the
context is atomic before obtaining a lock. Otherwise, if a spinlock is held while
you call lock/unclock, the scheduling might get corrupted.

Signed-off-by: Rubin Gerritsen <Rubin.Gerritsen@nordicsemi.no>

### CODE REVIEW

**Comment by rugeGerritsen** [comment]
@pdunaj, please take a look

File: drivers/bt_ll_nrfxlib/clock_control/clock_control_ll_nrfxlib.c [hunk 1]
```diff
@@ -16,6 +17,13 @@
 #include <hal/nrf_power.h>
 #endif
 
+static bool is_context_atomic(void)
+{
+	int key = irq_lock();
+	bool is_atomic = !z_arch_irq_unlocked(key) || z_is_in_isr();
+	irq_unlock(key);
+}
```
Comment (anangl) [inline drivers/bt_ll_nrfxlib/clock_control/clock_control_ll_nrfxlib.c]: Shouldn't `is_atomic` be returned here?

Comment (rugeGerritsen) [reply drivers/bt_ll_nrfxlib/clock_control/clock_control_ll_nrfxlib.c]: My bad. I'm surprised that CI tests did not catch this

**Comment by anangl** [comment]
Is this actually still needed? Haven't [this commit](https://github.com/NordicPlayground/fw-nrfconnect-zephyr/commit/312b43f14510754f18c8c6cf2cbabbd23f26484f#diff-23bba5938c8ac686f219f01f58478b9a) fixed this behavior?

**Comment by pdunaj** [comment]
Hi @anangl , the commit you mentioned fixed similar bug in the scheduler. The bug was caused by code entering the path for non-atomic context while context was actually atomic.

Here if check fails to detect that context is atomic code will cause lock to be acquired with K_FOREVER timeout. Code was detecting if we were in the interrupt but was not detecting if interrupts were disabled (e.g. if function was called under a spinlock or with irq_lock held).

**Comment by pdunaj** [comment]
@anangl , @rugeGerritsen BTW, there is another possible bug here. Lock is realized with a semaphore while it should be with a mutex. I mentioned it in the original PR introducing this code. Semaphores does not fix the priority inversion issue and so generally should not be used for locking.

**Comment by rugeGerritsen** [comment]
@pdunaj, you are right about that you pointed out that you preferred to use a mutex instead of a semaphore: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/687#discussion_r283328657. The reason for choosing to use a semaphore is that mutexes cannot be used in ISRs, see https://docs.zephyrproject.org/latest/reference/kernel/synchronization/mutexes.html. 
As far as I know, the fact that semaphores do not implement priority inheritance is does not prevent us from using the drivers.

**Comment by pdunaj** [comment]
> @pdunaj, you are right about that you pointed out that you preferred to use a mutex instead of a semaphore: [#687 (comment)](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/687#discussion_r283328657). The reason for choosing to use a semaphore is that mutexes cannot be used in ISRs, see https://docs.zephyrproject.org/latest/reference/kernel/synchronization/mutexes.html.
> As far as I know, the fact that semaphores do not implement priority inheritance is does not prevent us from using the drivers.

Yes, this is true and is the other side of the same coin. Since mutexes keep track of the owner they obviously cannot be used in the ISRs. I missed it... Thanks for pointing this out :)

File: drivers/clock_control/clock_control_nrfxlib.c [hunk 2]
```diff
@@ -29,9 +41,9 @@ static int hf_clock_start(struct device *dev, clock_control_subsys_t sub_system)
 
 	bool blocking = POINTER_TO_UINT(sub_system);
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply drivers/clock_control/clock_control_nrfxlib.c]: I agree this looks a bit odd. As far as I can see, the interpretation of `sub_system` is vendor specifc. For the clock driver in zephyr, this is not used. Does that mean that we should remove the blocking implementation?

**Comment by rugeGerritsen** [comment]
> The commit message should be changed, to adequately explain the motivation of the changes. The current one is misleading. The problem here is that when the code is executed in atomic context, it cannot perform any sleeping, so it cannot wait when acquiring the lock (I think it would be good to also mention it in the code, not only in the commit message). Referring the zephyrproject-rtos/zephyr#16273 issue does not really help in this context.
> The two other things signaled below are perhaps also worth considering.

@pdunaj , you reported this. Can you propose a good commit message for this?

**Comment by pdunaj** [comment]
This should be right:
```
Change is adding a check to ensure that when lock
is taken, function will not sleep while in an atomic context.
Atomic context can be an interrupt but also a thread
from which interrupts are disabled.
```

BTW. Let me check if performance of mouse is affected anyhow.

**Pull request merged**